feat(): type definition for signale

rjoydip opened the pull request:
Type definition for Signal.

Cludch commented:
+1
Hopefully this gets merged as I could use the typings currently.

jordrake commented on index.d.ts:
```diff
@@ -0,0 +1,126 @@
+export interface Stream { }
+export interface Config { displayScope?: Boolean; displayBadge?: Boolean; displayDate?: Boolean; displayFilename?: Boolean; displayLabel?: Boolean; displayTimestamp?: Boolean; underlineLabel?: Boolean; underlineMessage?: Boolean; underlinePrefix?: Boolean; underlineSuffix?: Boolean; uppercaseLabel?: Boolean; }
+export interface Options { times?: Times; types: Types; stream?: Stream; scope?: String; }
+export interface CompleteArrgumentType { prefix?: String; message: String; suffix?: String; }
```
You have misspelt argument. Unless you're a pirate ☠️

rjoydip replied:
@jordrake Can you tell me which one I misspelled?

Cludch replied:
`CompleteArrgumentType`

rjoydip replied:
@Cludch and @jordrake Thanks for pointing the issue to me. I fixed this.

resir014 commented:
Hey @rjoydip @klauscfhq,

As mentioned on #39, I accidentally started work on [my own version of the type declarations](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26411) without checking here if someone had done it first. It's currently awaiting PR on DefinitelyTyped, but I can port them over here and I'd be glad to contribute to this PR if needed. :)

Thanks!

rjoydip commented:
@resir014 Your type declarations is far better than my one. So if you give me the permission then I will take your code in here. Once it merged then anybody can extend. Otherwise, I will close this and you can make new PR.

resir014 commented:
@rjoydip Sure, no problem :) And feel free to add your name + github to the header as well.

rjoydip commented:
@resir014 Thank you so much.

resir014 commented:
@rjoydip No problem. Let me know if you have any questions.

rjoydip commented:
@resir014 added `Map` and `Set` interface. Also changed test case see below 
```js
import { Signale, SignaleOptions } from "../";

// --- Test 1: Basic Usage --- //
const signale =  new Signale();
```
I have removed `signale` as default export for avoid `error TS1192: Module has no default export` for the time being.

resir014 commented on signale.d.ts:
```diff
@@ -0,0 +1,148 @@
+    DefaultMethods: Signale.DefaultMethods;
+};
+
+declare function signale<T>(opt?: Signale.SignaleOptions): Signale.Signale<Signale.DefaultMethods> & {
```
Is this `declare function` necessary, btw? Shouldn't the above `declare const` solve the commonjs default export issue?

rjoydip replied:
I was trying to solve the `common.js default export issue` by this. I was following `fastify` definition file. If you say I could remove that.

resir014 commented on signale.d.ts:
```diff
@@ -0,0 +1,148 @@
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.7
+
+/// <reference types="node" />
```
This line could be removed, and we could add `@types/node` as a dependency/peerDependency.

rjoydip replied:
> including /// <reference types=" node" /> in a declaration file declares that this file uses names declared in @types/node/index.d.ts; and thus, this package needs to be included in the compilation along with the declaration file

Above statement has been taken from `typescript docs`. As per I understand if we keep this will haven't any problem in `compilation` time.

resir014 replied:
I see, feel free to keep this line then.

resir014 commented on signale.d.ts:
```diff
@@ -0,0 +1,148 @@
+/// <reference types="node" />
+
+interface Map<v, k> {}
+interface Set<T> {}
```
These interfaces should be included in the TS stdlib, so they can be removed.

rjoydip replied:
TS lib not including those. I have tried a couple of solutions and lastly, I got this solution [here](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/8264#issuecomment-189247585).

resir014 replied:
I think we can still bump `"lib"` up to `"es6"` inside the tsconfig. That way both `Map` and `Set` are included in the stdlib, and these two lines could be removed.

Besides, the library is already based on es6 classes anyway :)

rjoydip replied:
I had tried too.

rjoydip replied:
![capture](https://user-images.githubusercontent.com/15318294/41202917-fdfa3a36-6cef-11e8-929d-35a4de9a5a65.PNG)

resir014 replied:
Actually, disregard my last comment. I just cloned your branch locally and just saw what you mean. It's safe to still keep these two declaration there.

Ignore my last comment, thanks. :)

resir014 commented on test/signale-tests.js:
```diff
@@ -0,0 +1,101 @@
+"use strict";
```
This file should be removed now that it's ignored.

rjoydip replied:
removed this one also

resir014 commented on test/tslint.json:
```diff
@@ -0,0 +1,3 @@
+{
```
This file is also only necessary for DefinitelyTyped, so you can remove this.

rjoydip replied:
removed this one

resir014 commented:
Thanks @rjoydip!

> I have removed `signale` as default export for avoid error TS1192: Module has no default export for the time being.

Alright, I think I can take over on a separate PR to try and fix this issue over the next few days. For now I think we can merge this.

rjoydip commented:
Awesome.

rjoydip commented:
Yes. @resir014 Thanks once again. It couldn't be good without your help.

rjoydip commented:
As like this issue #42. When i was checking `signale.success(Symbol('test').toString());` I am getting this error `error TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'SymbolConstructor' has no compatible call signatures.`

have to check this one

resir014 commented:
@rjoydip Alright. If you want me to, you can also give me push access to your fork and I can work on it as well. :)

resir014 commented:
Thanks @rjoydip!

I'm going to make a few changes:

* I've renamed the `tests` folder to the `types` folder, and moved the `signale.d.ts` file in there.
* This means we can include the `.d.ts` file inside the `types` folder and not worry about the compiler errors on the `Map` interface.
* Attempted to fix the commonjs default export issue

rjoydip commented:
No problem. Change it

resir014 commented:
It's done. :)

On Mon, 11 Jun 2018, 12:43 Joydip Roy, <notifications@github.com> wrote:

> No problem. Change it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/klauscfhq/signale/pull/25#issuecomment-396128264>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFZsha5OjDET77iS5wOAD3LXVCP-nL-5ks5t7gOUgaJpZM4UPWTH>
> .
>

kingdaro commented on types/signale.d.ts:
```diff
@@ -0,0 +1,134 @@
+};
+
+export = signale;
+export as namespace signale;
```
The `export as namespace` declaration is used with modules that export some sort of global variable. See [here](https://stackoverflow.com/a/44849515/1332403) and [here](https://github.com/zspitz/TypeScript-Handbook/blob/release-2.0/pages/Modules.md#umd-modules). Since Signale appears to be a CommonJS-only module and doesn't export a global, this line can probably be removed.

resir014 replied:
Fixed. 👌

resir014 commented:
@klauscfhq this should be ready to review 👌

Eke commented:
Hello, any updates on this? I would love to have this in my project.

resir014 commented:
@Eke Hi, I'm still pushing for this to be merged as well. I've updated the typings with the missing config objects from version 1.2.0 and now this should be ready for @klauscfhq to review and merge.
